sklearn/neural_network/_base.py

```diff
@@ -177,6 +189,33 @@ def squared_loss(y_true, y_pred, sample_weight=None):
     )
 
 
+def poisson_loss(y_true, y_pred, sample_weight=None):
+    """Compute (half of the) Poisson deviance loss for regression.
+
+    Parameters
+    ----------
+    y_true : array-like or label indicator matrix
+        Ground truth (correct) labels.
+
+    y_pred : array-like or label indicator matrix
+        Predicted values, as returned by a regression estimator.
+
+    sample_weight : array-like of shape (n_samples,), default=None
+        Sample weights.
+
+    Returns
+    -------
+    loss : float
+        The degree to which the samples are correctly predicted.
+    """
+    # TODO: Decide what to do with the term `xlogy(y_true, y_true) - y_true`. For now,
+    # it is included. But the _loss module doesn't use it (for performance reasons) and
+    # only adds it as return of constant_to_optimal_zero (mainly for testing).
```

Comment:
This also needs a decision.

Reply:
> But the _loss module doesn't use it (for performance reasons)

Is there any real performance impact of computing this term when training a model?

I think I would favor always including it. If it used for monitoring, displaying something named "Poisson loss" which is actually not the true loss, but the loss with a constant offset can be quite surprising.

Reply:
Alternatively, we could just rename this function to `offset_poisson_loss` to make it explicit that the computation of the constant terms are removed.

But if we expose the value of the objective function under the name "Poisson loss" via the public API, one way or another, I think we should make sure we re-add the constant terms to avoid any confusion when comparing values on resampled data (e.g. between train and test split to check for overfitting).

Reply:
I don't think optimization is too much of a concern with this module so we might as well keep the constant terms.

Reply:
> I don't think optimization is too much of a concern with this module so we might as well keep the constant terms.

The constant term in the loss only matters if we report/save those loss values as we do in MLP (and in GradientBoosting, but not in HistGradientBoosting). The nice thing about adding the constant term is that the minimum of the loss is then 0 exactly. Omitting it saves a bit of time.
Note also that NN libraries like pytorch usually omit the constant term (see https://pytorch.org/docs/stable/generated/torch.nn.PoissonNLLLoss.html#torch.nn.PoissonNLLLoss).

Reply:
Yes Tensorflow also omits it https://www.tensorflow.org/api_docs/python/tf/keras/losses/Poisson.

I would say let's keep it for now. Would we accomplish much by saving time in this module when there are other optimal libraries just for implementing neural networks?